OSTimer updates

felipebalbi opened the pull request:
Setup ostimer interrupts internally such that users always get a working tick.

Commit:
Initialize OSTIMER0 during HAL initialization

Provide the user with a working time driver.

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

Commit:
Handle time_driver interrupt internally

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

jamesmunns commented on examples/src/bin/ostimer_race_test.rs:
```diff
@@ -96,12 +87,6 @@ async fn main(_spawner: Spawner) {
-
     uart.write_str_blocking("Time driver initialized\n");
 
     // Create OSTIMER instance
```
Do we want to feature-gate time-driver support, and avoid giving the user the OSTIMER peripheral when it is enabled? This code is probably Not Good since the time-driver is also active at the same time

jamesmunns replied:
https://github.com/OpenDevicePartnership/embassy-mcxa/blob/main/src/lib.rs#L86 specifically is probably what we should gate if a `time` feature is active. This somewhat matches what embassy-stm32 does, though we only have one time driver impl currently (we might want an RTC based one later):

https://github.com/embassy-rs/embassy/blob/a72ba1e0571e6bef9c93c1d98b67724c685a37cc/embassy-stm32/Cargo.toml#L265-L301

jamesmunns commented:
Following up with https://github.com/OpenDevicePartnership/embassy-mcxa/pull/24#discussion_r2539629880, it's maybe a little more complicated. Some notes:

* We can't exactly JUST gate the peripheral, without reworking how the `src/ostimer.rs` file is shaped. The public `Ostimer` API is also used by the `time-driver` impl, so we probably need to make it somewhat private?
* But also, the current `time-driver` impl requires the `OSTIMER` peripheral to call the `clocks::enable_and_reset` function. We could work around this, but it's a bit annoying.
* While `embassy-stm32` does some complicated codegen to notch out the timer used by `time-driver`, other HALs like `embassy-rp` just don't provide a public API for timers at all? So we could potentially not have a public API for `Ostimer`, for now at least.

Also, as currently written, the `ostimer_*` examples are *probably* not ideal: They manually create an `Ostimer` instance, while also assume that the `time-driver` (which also drives the ostimer!) is active, mixing explicit calls with things like `Timer::after(...).await`. We might want to remove these, or split them into a separate project that doesn't activate the time-driver feature (separate features for separate bins in the same project isn't great).

I'm going to make a PR to this PR that just does these things, but we may want to think about whether to merge it or not.

jamesmunns committed:
Gate `time-driver` impl behind a `time` flag

Also prevents creation of an `Ostimer` instance if the `time`
feature is active.

jamesmunns committed:
Remove some dead code

felipebalbi committed:
Merge pull request #1 from jamesmunns/james/more-ostimer

Gate `time-driver` impl behind a `time` flag
